src/cmd/serve.rs

```diff
@@ -148,6 +165,61 @@ async fn handle_request(req: Request<Body>, mut root: PathBuf) -> Result<Respons
         .unwrap())
 }
 
+/// Inserts build error message boxes into HTML responses when needed.
+async fn response_error_injector(
+    req: impl IntoFuture<Output = Result<Response<Body>>>,
+) -> Result<Response<Body>> {
+    let req = req.await;
+
+    // return req as-is if the request is Err(), not HTML, or if there are no error messages.
+    if req
+        .as_ref()
+        .map(|req| {
+            req.headers()
+                .get(header::CONTENT_TYPE)
+                .map(|val| val != &HeaderValue::from_static("text/html"))
+                .unwrap_or(true)
+        })
+        .unwrap_or(true)
+        || SERVE_ERROR.lock().unwrap().get_mut().is_none()
+    {
+        return req;
+    }
+
+    let mut req = req.unwrap();
+    let mut bytes = body::to_bytes(req.body_mut()).await.unwrap().to_vec();
+
+    if let Some((msg, error)) = SERVE_ERROR.lock().unwrap().get_mut() {
+        // Generate an error message similar to the CLI version in messages::unravel_errors.
+        let mut error_str = String::new();
+
+        if !msg.is_empty() {
+            error_str.push_str(&format!("Error: {msg}\n"));
+        }
+
+        error_str.push_str(&format!("Error: {error}\n"));
+
+        let mut cause = error.source();
+        while let Some(e) = cause {
+            error_str.push_str(&format!("Reason: {}\n", e));
+            cause = e.source();
+        }
+
+        // Push the error message (wrapped in an HTML dialog box) to the end of the HTML body.
+        //
+        // The message will be outside of <html> and <body> but web browsers are flexible enough
+        // that they will move it to the end of <body> at page load.
+        let html_error = format!(
+            r#"<div style="all:revert;position:fixed;display:flex;align-items:center;justify-content:center;background-color:rgb(0,0,0,0.5);top:0;right:0;bottom:0;left:0;"><div style="background-color:white;padding:0.5rem;border-radius:0.375rem;filter:drop-shadow(0,25px,25px,rgb(0,0,0/0.15));overflow-x:auto;"><p style="font-weight:700;color:black;font-size:1.25rem;margin:0;margin-bottom:0.5rem;">Zola Build Error:</p><pre style="padding:0.5rem;margin:0;border-radius:0.375rem;background-color:#363636;color:#CE4A2F;font-weight:700;">{error_str}</pre></div></div>"#
+        );
+        bytes.extend(html_error.as_bytes());
```

Comment:
Can we get some invalid HTML generated by inserting after `</html>`?

Reply:
Yes, this way is technically invalid HTML, but it worked on all the browsers I tested it on. If that's too much of a problem I can try to remove `</html>` before appending the message box 🤔

Reply:
If it works on all browsers it's fine, not the weirdest HTML they will see